File: SideTableDemo.go
package main

import (
	"fmt"
)

type Expr struct {
	Name string
}

func main() {
	fmt.Println("=== Go Map Key Behavior Demo ===\n")

	// Scenario 1: Using Struct Values as Keys
	// In Go, structs are comparable by value (content).
	fmt.Println("--- Scenario 1: Struct Values as Keys ---")
	mapByValue := make(map[Expr]int)

	val1 := Expr{Name: "a"}
	val2 := Expr{Name: "a"}

	fmt.Printf("val1: %v\n", val1)
	fmt.Printf("val2: %v\n", val2)
	fmt.Printf("Are they equal? (val1 == val2): %v\n", val1 == val2)

	mapByValue[val1] = 1
	fmt.Println("Set map[val1] = 1")

	// This will OVERWRITE the previous entry because val1 == val2
	mapByValue[val2] = 2
	fmt.Println("Set map[val2] = 2")

	fmt.Printf("Value for val1: %d\n", mapByValue[val1]) // Will be 2
	fmt.Printf("Value for val2: %d\n", mapByValue[val2]) // Will be 2
	fmt.Println()

	// Scenario 2: Using Pointers as Keys
	// Pointers are comparable by address (reference identity).
	// This mimics the Java behavior we saw.
	fmt.Println("--- Scenario 2: Pointers as Keys ---")
	mapByPointer := make(map[*Expr]int)

	ptr1 := &Expr{Name: "a"}
	ptr2 := &Expr{Name: "a"}

	fmt.Printf("ptr1: %p (Name: %s)\n", ptr1, ptr1.Name)
	fmt.Printf("ptr2: %p (Name: %s)\n", ptr2, ptr2.Name)
	fmt.Printf("Are they equal? (ptr1 == ptr2): %v\n", ptr1 == ptr2)

	mapByPointer[ptr1] = 1
	fmt.Println("Set map[ptr1] = 1")

	// This will create a NEW entry because ptr1 != ptr2
	mapByPointer[ptr2] = 42
	fmt.Println("Set map[ptr2] = 42")

	fmt.Printf("Value for ptr1: %d\n", mapByPointer[ptr1]) // Will be 1
	fmt.Printf("Value for ptr2: %d\n", mapByPointer[ptr2]) // Will be 42
}

File: SideTableDemo.hs
module Main where

import Data.Map (Map)
import qualified Data.Map as Map
import Data.Unique -- This provides "Object Identity" capabilities

-- ==========================================
-- Approach 1: Pure Values (The Default)
-- ==========================================
-- Values are defined by their content.
newtype Expr = Var String deriving stock (Show, Eq, Ord)

demoPure :: IO ()
demoPure = do
  putStrLn "--- Approach 1: Pure Values (Default) ---"
  let val1 = Var "a"
  let val2 = Var "a"

  putStrLn $ "val1: " ++ show val1
  putStrLn $ "val2: " ++ show val2
  putStrLn $ "Are they equal? " ++ show (val1 == val2)

  let m = Map.fromList [(val1, 1)]
  -- This overwrites because val1 and val2 are structurally identical
  let m' = Map.insert val2 2 m

  putStrLn $ "Value for val1: " ++ show (Map.lookup val1 m')
  putStrLn $ "Value for val2: " ++ show (Map.lookup val2 m')
  putStrLn ""

-- ==========================================
-- Approach 2: Simulating References with Unique IDs
-- ==========================================
-- This is how we get "Reference Semantics" in Haskell.
-- We explicitly attach a unique identifier generated in IO.

data RefExpr = RefExpr Unique String

-- We define Equality and Ordering based ONLY on the Unique ID
instance Eq RefExpr where
  (RefExpr u1 _) == (RefExpr u2 _) = u1 == u2

instance Ord RefExpr where
  compare (RefExpr u1 _) (RefExpr u2 _) = compare u1 u2

instance Show RefExpr where
  show (RefExpr u s) = "RefExpr(" ++ show (hashUnique u) ++ ") " ++ show s

demoReference :: IO ()
demoReference = do
  putStrLn "--- Approach 2: Simulating References (Data.Unique) ---"

  -- We must create these in the IO monad, just like 'new' in Java
  id1 <- newUnique
  id2 <- newUnique

  let ref1 = RefExpr id1 "a"
  let ref2 = RefExpr id2 "a"

  putStrLn $ "ref1: " ++ show ref1
  putStrLn $ "ref2: " ++ show ref2
  putStrLn $ "Are they equal? " ++ show (ref1 == ref2)

  let m = Map.fromList [(ref1, 1)]
  -- This creates a NEW entry because id1 /= id2
  let m' = Map.insert ref2 42 m

  putStrLn $ "Value for ref1: " ++ show (Map.lookup ref1 m')
  putStrLn $ "Value for ref2: " ++ show (Map.lookup ref2 m')

main :: IO ()
main = do
  putStrLn "=== Haskell Reference Types Demo ===\n"
  demoPure
  demoReference
